### PR TITLE
Windows build fixes

### PR DESCRIPTION
**Pull request by dtasev**
- Added /bigobj for test executable, only for WIN32
- Extracted functions from DataArray `operator*`/`operator/` to enforce order. Issue seen was an unexpected exception being, due to different function call order between gcc/msvc
- Fixed losing value in transform_subspan::maybe_subspan by wrapping variable in `unique_ptr`
- Added size parameter to element_array to distinguish between a null and an empty array. This is necessary as otherwise both begin/end iterators point at `nullptr` when copying a null or when copying an empty array. However, a null has a size of -1, and an empty has a size of 0.

### CODE REVIEW

File: core/include/scipp/core/element_array.h [hunk 1]
```diff
@@ -77,7 +81,7 @@ template <class T> class element_array {
   }
 
   element_array(const element_array &other)
-      : element_array(other.data(), other.data() + other.size()) {}
+      : element_array(other.begin(), other.end(), other.size()) {}
```
Comment (igudich) [inline core/include/scipp/core/element_array.h]: I wouldn't mix the constructor for normal iterators let's `std::vector::iterator` with the case of copy constructor for `element_array`. It is better to introduce the private fabric function let say `make_element_array(const element_array & other)` and call it. Then to avoid usage of wrong constructor with iterators lets say this shouldn't compile:
```c++
element_array arr;
element_array arr1(arr.begin(), arr.end());
``` 
I would change signature to:
```
template <class Iter>
std::enable_if_t<std::is_assignable<T &, Iter::value_type)>{}, int> = 0>
element_array(Iter first, Iter last);
```

File: core/include/scipp/core/transform_subspan.h [hunk 2]
```diff
@@ -22,12 +22,12 @@ static constexpr auto need_subspan = [](const VariableConstProxy &var,
...
-  Variable ret;
+  auto ret = std::make_unique<Variable>();
   if (need_subspan(var, dim)) {
-    ret = subspan_view(var, dim);
-    var = ret;
+    *ret = subspan_view(var, dim);
+    var = *ret;
   }
-  return ret;
+  return std::move(ret);
```
Comment (igudich) [inline core/include/scipp/core/transform_subspan.h]: Is `keep_subspan_vars_alive` in line 52 below needed now?

File: core/include/scipp/core/except.h [hunk 3]
```diff
@@ -35,7 +35,10 @@ struct SCIPP_CORE_EXPORT TypeError : public std::runtime_error {
   using std::runtime_error::runtime_error;
 
   template <class... Vars>
-  TypeError(const std::string &msg, const Vars &... vars)
+  TypeError(const std::string &msg) : std::runtime_error(msg) {}
```
Comment (igudich) [inline core/include/scipp/core/except.h]: Why is this needed?

File: core/include/scipp/core/transform_subspan.h (hunk 2 above)
Comment (dtasev) [reply core/include/scipp/core/transform_subspan.h]: I think so, it's the one that will keep/destroy the `unique_ptr`s in that scope

File: core/include/scipp/core/except.h (hunk 3 above)
Comment (dtasev) [reply core/include/scipp/core/except.h]: Some TypeError calls do not have anything in the argpack. I was getting a compile error without this one.

File: core/include/scipp/core/element_array.h (hunk 1 above)
Comment (dtasev) [reply core/include/scipp/core/element_array.h]: > It is better to introduce the private fabric function let say make_element_array(const element_array & other) and call it

I don't understand where that function would be called: in a new copy constructor, or whenever we want to do 

```
element_array arr;
element_array arr1(arr.begin(), arr.end());
```

we would instead have to do

```
element_array arr;
element_array arr1 = arr.make_element_array()
```

Comment (igudich) [reply core/include/scipp/core/element_array.h]: > > It is better to introduce the private fabric function let say make_element_array(const element_array & other) and call it
> 
> I don't understand where that function would be called: in a new copy constructor, or whenever we want to do
> 
> ```
> element_array arr;
> element_array arr1(arr.begin(), arr.end());
> ```
> 
> we would instead have to do
> 
> ```
> element_array arr;
> element_array arr1 = arr.make_element_array(arr)
> ```

But now you just expose  the broken API, I'm a user and I can write like this:
```c++
element_array arr;
element_array arr1(arr.begin(), arr.end());
```
and have no compilation error, but the wrong result.
I'm suggesting:
```c++
private:
  static element_array from_other(const element_array& arr) {
    if (arr.size() == -1)
      return element_array();
    if (arr.size() = 0)
       return element_array(0);
    else ....
  }
public:
  element_array(const element_array& other) : element_array(from_other(other));
```

File: core/include/scipp/core/element_array.h [hunk 4]
```diff
@@ -53,9 +54,12 @@ template <class T> class element_array {
       class Iter,
       std::enable_if_t<
           std::is_assignable<T &, decltype(*std::declval<Iter>())>{}, int> = 0>
-  element_array(Iter first, Iter last) {
-    resize(std::distance(first, last), default_init_elements);
-    std::copy(first, last, data());
+  element_array(Iter first, Iter last,
+                std::optional<scipp::index> size = std::nullopt) {
+    if (!size.has_value() || (size.has_value() && *size > -1)) {
```
Comment (OwenArnold) [inline core/include/scipp/core/element_array.h]: I think it opening up  another issue to deal with avoiding coding against these special meaning integer values would be good. Maybe define some named `static const` variables to give better meaning to these.

**Comment by SimonHeybrock** [comment]
>     * Added size parameter to element_array to distinguish between a null and an empty array. This is necessary as otherwise both begin/end iterators point at `nullptr` when copying a null or when copying an empty array. However, a null has a size of -1, and an empty has a size of 0.

Can you explain what issue this caused?

File: core/include/scipp/core/element_array.h (hunk 4 above)
Comment (SimonHeybrock) [reply core/include/scipp/core/element_array.h]: I don't think we should change this constructor. Just fix the copy constructor below.

Or is there another issue where this is relevant?

**Comment by dtasev** [comment]
Copying an empty array causes the [copy constructor](https://github.com/scipp/scipp/blob/master/core/include/scipp/core/element_array.h#L56-L59) `std::distance(nullptr, nullptr - 1)` to give a nonsense value. `nullptr - 1` on MSVC is some value. The actual test failure happens afterwards: because that value is rarely going to be -1, then the bool/empty check fails.

**Comment by SimonHeybrock** [comment]
> Copying an empty array causes the [copy constructor](https://github.com/scipp/scipp/blob/master/core/include/scipp/core/element_array.h#L56-L59) to essentially calculate `std::distance(nullptr, nullptr - 1)` which gives a nonsense value. `nullptr - 1` on MSVC is some value. The actual test failure happens afterwards: because that value is rarely going to be -1, then the bool/empty check fails.

As mentioned in another comment, I suggest to fix the copy constructor, not change the constructor from iterators.

File: core/include/scipp/core/element_array.h (hunk 4 above)
Comment (dtasev) [reply core/include/scipp/core/element_array.h]: The only issue is then copying via iterators will have a different result than via the copy constructor in the output element array.

Comment (SimonHeybrock) [reply core/include/scipp/core/element_array.h]: Well, it will also as you have it now, with the default argument.

I agree it is not beautiful, but I do not see it as an issue, since this is for internal use with a very limited number of call sites.

File: core/include/scipp/core/element_array.h [hunk 5]
```diff
@@ -7,6 +7,7 @@
 
 #include <algorithm>
 #include <memory>
+#include <optional>
```
Comment (SimonHeybrock) [inline core/include/scipp/core/element_array.h]: Can remove this now?

**Comment by dtasev** [comment]
I've added the changes for the `element_array` copy constructor. Are we implementing another workaround instead of the `unique_ptr` in this PR?

**Comment by SimonHeybrock** [comment]
> I've added the changes for the `element_array` copy constructor. Are we implementing another workaround instead of the `unique_ptr` in this PR?

I don't particularly like the workaround, but I am fine with leaving it. :)
